File: problem_016.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "swiftc problem_016.swift -o problem_016" -*-
// Copyright (c) 2016 Michael Caldwell

print("Project Euler - Problem 16:")
print("What is the sum of the digits of the number 2^1000\n")

var num: [Int] = [2]
var debug: Bool = false

for i in 2...1000 {
    for n in 0...num.count-1 {
        num[n] *= 2
    }

    for n in 0...num.count-1 {
        while num[n] > 9 {
            num[n] -= 10
            if n == num.count-1 {
                num.append(0)
            }
            num[n+1] += 1
        }
    }
}

var sum = 0
for var n in stride(from: num.count-1, to: 0, by: -1) {
    sum += num[n]
    if debug == true {
        print(num[n], terminator: "")
    }
}

print(sum)

File: problem_016.rs
// -*- compile-command: "rustc -o problem_016_rs problem_016.rs" -*-
// Copyright (c) 2016 Michael Caldwell


fn main() {
    println!("Project Euler - Problem 16");
    println!("What is the sum of the digits of the number 2^1000\n");

    let mut bignum = Vec::new();
    bignum.push(1);

    for _ in 0..1000 {
        // Multiple all elements by 2
        for i in 0..bignum.len() {
            bignum[i] *= 2;
        }

        // Carry cells over to the right
        for i in 0..bignum.len() {
            while bignum[i] > 9 {
                if i+1 >= bignum.len() {
                    bignum.push(0);
                }
                bignum[i] -= 10;
                bignum[i+1] += 1;
            }
        }
    }

    let mut sum = 0;
    for i in 0..bignum.len() {
        sum += bignum[i];
    }

    println!("{}", sum);
}
